AsyncDisplayKit/Layout/ASLayoutOptions.mm

```diff
@@ -0,0 +1,252 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import "ASLayoutOptions.h"
+
+#import <AsyncDisplayKit/ASAssert.h>
+#import <AsyncDisplayKit/ASThread.h>
+#import <AsyncDisplayKit/ASTextNode.h>
+#import "ASInternalHelpers.h"
+
+@interface ASLayoutOptions()
+{
+  ASDN::RecursiveMutex _propertyLock;
+}
+@end
+
+@implementation ASLayoutOptions
+
+@synthesize spacingBefore = _spacingBefore;
+@synthesize spacingAfter = _spacingAfter;
+@synthesize flexGrow = _flexGrow;
+@synthesize flexShrink = _flexShrink;
+@synthesize flexBasis = _flexBasis;
+@synthesize alignSelf = _alignSelf;
+
+@synthesize ascender = _ascender;
+@synthesize descender = _descender;
+
+@synthesize sizeRange = _sizeRange;
+@synthesize layoutPosition = _layoutPosition;
+
+static Class gDefaultLayoutOptionsClass = nil;
++ (void)setDefaultLayoutOptionsClass:(Class)defaultLayoutOptionsClass
+{
+  gDefaultLayoutOptionsClass = defaultLayoutOptionsClass;
+}
+
++ (Class)defaultLayoutOptionsClass
+{
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    if (gDefaultLayoutOptionsClass == nil) {
+      // If someone is asking for this and it hasn't been customized yet, use the default.
+      gDefaultLayoutOptionsClass = [ASLayoutOptions class];
+    }
+  });
+  return gDefaultLayoutOptionsClass;
+}
+
+- (instancetype)init
+{
+  return [self initWithLayoutable:nil];
+}
+
+- (instancetype)initWithLayoutable:(id<ASLayoutable>)layoutable;
+{
+  self = [super init];
+  if (self) {
+    [self setupDefaults];
+    [self setValuesFromLayoutable:layoutable];
+  }
+  return self;
+}
+
+#pragma mark - NSCopying
+- (id)copyWithZone:(NSZone *)zone
+{
+  ASLayoutOptions *copy = [[[self class] alloc] init];
+  [self copyIntoOptions:copy];
+  return copy;
+}
+
+- (void)copyIntoOptions:(ASLayoutOptions *)copy
+{
+  ASDN::MutexLocker l(_propertyLock);
+  copy.flexBasis = self.flexBasis;
+  copy.spacingAfter = self.spacingAfter;
+  copy.spacingBefore = self.spacingBefore;
+  copy.flexGrow = self.flexGrow;
+  copy.flexShrink = self.flexShrink;
+  
+  copy.ascender = self.ascender;
+  copy.descender = self.descender;
+  
+  copy.sizeRange = self.sizeRange;
+  copy.layoutPosition = self.layoutPosition;
+}
+
+
+#pragma mark - Defaults
+- (void)setupDefaults
```

Comment:
I'm thinking we should kill this method and have people do their configuration in -init.  A subclass will likely override -init out of habit anyway, and would naturally put initialization of any new data structures there, so I'm not sure there is value in -setupDefaults.

More importantly, let's comment out any of these lines that set already-default values.  Since this object will be created / very / frequently, and currently will be created on the main thread, we need to cut down on method calls; there are at least 8 below that will always match the calloc / zeroed-out state that Objective-C objects are created with.  I'm not sure how the enum values are defined, but assume they are also equal to 0.

So you could add a comment explaining that, and keep the lines (commented out) for explicitness if desired, or just remove them.